ggml-cuda.cu

```diff
@@ -5987,6 +6035,527 @@ static  __global__ void im2col_f32_f16(
+                            half16x16_a mv;
+                            nvcuda::wmma::load_matrix_sync(mv, ss + 16*j*T + 16*cc, T);
+
+                            nvcuda::wmma::mma_sync(lo[j][i], mv, mk, lo[j][i]);
+                        }
+                    }
+                }
+            }
+        }
+
+        // these are needed for reducing the results from the simdgroups (reuse the ss buffer)
+        for (int64_t j = 0; j < Q; ++j) {
+            if (lane_id == 0) {
+                ss[j*T + 0] = __float2half(S[j]);
+                ss[j*T + 1] = __float2half(M[j]);
+            }
+        }
+    }
+
+    // reduce the warps sequentially
+    for (int64_t sg = 1; sg < num_warps; ++sg) {
+        float S = 0.0f;
+        float M = -INFINITY;
+
+        __syncthreads();
+
+        // each simdgroup stores its output to shared memory, reusing sq
+        if (warp_id == sg) {
+            for (int64_t j = 0; j < Q16; ++j) {
+                for (int64_t i = 0; i < D16; ++i) {
+                    nvcuda::wmma::store_matrix_sync(sq + 16*j*T + i*16, lo[j][i], T, nvcuda::wmma::mem_col_major);
+                }
+            }
+        }
+
+        __syncthreads();
+
+        // the first simdgroup accumulates the results from the other simdgroups
+        if (warp_id == 0) {
+            for (int64_t j = 0; j < Q; ++j) {
+                const float S0 = __half2float(ss[j*T +         0]);
+                const float S1 = __half2float(ss[j*T + sg*SH + 0]);
+
+                const float M0 = __half2float(ss[j*T +         1]);
+                const float M1 = __half2float(ss[j*T + sg*SH + 1]);
+
+                M = max(M0, M1);
+
+                const float ms0 = M0 == -INFINITY ? 0.0f : expf(M0 - M);
+                const float ms1 = M1 == -INFINITY ? 0.0f : expf(M1 - M);
+
+                S = S0*ms0 + S1*ms1;
+
+                if (lane_id == 0) {
+                    ss[j*T + 0] = __float2half(S);
+                    ss[j*T + 1] = __float2half(M);
+
+                    ss[j*T + C + j        ] = __float2half(ms0);
+                    ss[j*T + C + j + sg*SH] = __float2half(ms1);
+                }
+            }
+
+            // O_0 = diag(ms0)*O_0 + diag(ms1)*O_1
+            for (int64_t j = 0; j < Q16; ++j) {
+                half16x16_a ms0;
+                half16x16_a ms1;
+                half16x16_b t;
+                half16x16_acc t2;
+
+                nvcuda::wmma::load_matrix_sync(ms0, ss + 16*j*T + C + 16*j,         T);
+                nvcuda::wmma::load_matrix_sync(ms1, ss + 16*j*T + C + 16*j + sg*SH, T);
+
+                for (int64_t i = 0; i < D16; ++i) {
+                    nvcuda::wmma::load_matrix_sync(t, sq + 16*j*T + i*16, T);
+                    nvcuda::wmma::mma_sync(t2, ms1, t, t2);
+
+                    // t <- lo
+                    for (uint32_t k = 0; k < t.num_elements; k++) {
+                        t.x[k] = lo[j][i].x[k];
+                    }
```

Comment:
NVIDIA does **not** give you any guarantees regarding how the data for tensor core fragments is distributed between threads in a warp. So even if the memory layout between the two matrix types happens to be the same on the GPUs that you're testing your code on it may produce incorrect results on other architectures.

For now I would recommend that you replace this with `store_matrix_sync` and `load_matrix_sync` to make sure it's not the reason the results are incorrect. Maybe check the FlashAttention CUDA code to see what they do.